packages/ai-jsx/src/lib/tts/tts.tsx

```diff
@@ -0,0 +1,360 @@
+   */
+  public onComplete?: () => void;
+
+  constructor(protected readonly name: string) {
+    this.audio = new Audio();
+    this.audio.onplay = () => console.log(`[${this.name}] tts playing`);
+    this.audio.onpause = () => console.log(`[${this.name}] tts paused`);
+    this.audio.onloadstart = () => console.log(`[${this.name}] tts loadstart`);
+    this.audio.onloadeddata = () => console.log(`[${this.name}] tts loadeddata`);
+    this.audio.oncanplay = () => console.log(`[${this.name}] tts canplay`);
+    this.audio.onplaying = () => {
+      console.log(`[${this.name}] tts playing`);
+      if (this.playMillis) {
+        this.latency = Math.floor(performance.now() - this.playMillis);
+        console.log(`[${this.name}] tts play latency: ${this.latency} ms`);
+      }
+      this.onPlaying?.();
+    };
+  }
+  /**
+   * Converts the given text to speech and plays it using the HTML5 audio element.
+   * This method may be called multiple times, and the audio will be played serially.
+   * Generation may be buffered, use the flush() method to indicate all text has been provided.
+   */
+  play(_text: string) {
+    throw new Error('Method not implemented.');
+  }
+  /**
+   * Flush any buffered text and play the audio.
+   */
+  flush() {}
+  /**
+   * Stop playing any generated audio and discard any buffered text.
+   */
+  stop() {
+    console.log(`[${this.name}] tts stopping`);
+    this.audio.pause();
+    this.cleanUp();
+  }
+  protected cleanUp() {}
+}
+
+/**
+ * Defines a text-to-speech service that requests audio from a server and
+ * plays it in one shot using the HTML5 audio element. The URL to request
+ * audio from the server is constructed using the provided BuildUrlFunction,
+ * allowing this class to be used with a variety of text-to-speech services.
+ */
+export class SimpleTextToSpeech extends TextToSpeechBase {
+  constructor(
+    name: string,
+    protected readonly urlFunc: BuildUrlFunction,
+    private readonly voice: string,
+    private readonly rate: number = 1.0
+  ) {
+    super(name);
+  }
+  play(text: string) {
+    this.playMillis = performance.now();
+    this.audio.src = this.urlFunc(this.name, this.voice, this.rate, text);
+    this.audio.play();
+  }
+}
+
+class AudioChunk {
+  public timestamp: number;
+  constructor(public buffer?: ArrayBuffer) {
+    this.timestamp = performance.now();
+  }
+}
+
+/**
+ * Defines a text-to-speech service that requests individual audio utterances
+ * from a server and plays them in series using HTML5's Media Source Extensions.
+ * This approach reduces latency by allowing the audio to be streamed as it is
+ * generated, rather than waiting for the entire audio file to be generated. It also
+ * allows text to be fed to the service in a stream rather than all at once. This
+ * class is not meant to be used directly, but provides infrastructure for
+ * RestTextToSpeech and ElevenLabsTextToSpeech.
+ */
+export class MseTextToSpeech extends TextToSpeechBase {
```

Comment:
If it's not meant to be used directly, maybe we don't want to `export` it? It can also be marked as `@hidden` to prevent it from showing up in the auto-generated API docs.